### DISCUSSION TITLE
CORS Middleware: trim trailing space in allowed_origins

### DISCUSSION THREAD

**Discussion by sultaniman** [encode/starlette#2735] [Potential Issue] [answered]
Hey all,

At the moment the middleware is unable to match origin if the given origins contains trailing slash.

```py
app.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost:5173/"]    # trailing /
    ...
)
```

When the browser sends the `Origin` header the `CORSMiddleware.is_allowed_origin` is not able to match it

```
Host: localhost:8000
Origin: http://localhost:5173
```

https://github.com/encode/starlette/blob/master/starlette/middleware/cors.py#L95-L102

I know it is a half measure but still wondering if we need to trim trailing slash from `allowed_origins`?

**Comment by scy** [comment]
I've just been bitten by this issue.

However, instead of silently trimming the slash, my suggestion would be that `CORSMiddleware` raises an exception if one of the `allowed_origins` doesn't look like a valid origin value (e.g. because it's ending in a slash). That way, the user actually notices that something is wrong with the value they're passing, and is given the chance (and obligation) to fix it.

PS: This discussion's subject says "trim trailing space", but that should probably be "trim trailing slash" instead.

**Reply by Kludex** [reply]
Which kind of exception?

**Reply by scy** [reply]
How about a simple `ValueError`?

I'd implement it like this: In the `CORSMiddleware` constructor, if `allowed_origins` is set, iterate over it and check whether each value is either `*` or matches `r"^[a-zA-Z][a-zA-Z0-9.+-]*://[][a-zA-Z0-9._~%!$&'()*+,;=-]*(?::[0-9]+)?$"`, which is how RFC 6454 (in combination with RFC 3986) defines a origin. If the value doesn't conform, `raise ValueError(f"invalid CORS origin: {item}")`.

**Reply by Kludex** [reply]
I don't think a 500 should happen here.

**Reply by scy** [reply]
Why a 500? If it's implemented in `CORSMiddleware.__init__`, the exception is raised on application startup, not on a request, isn't it?

**Reply by Kludex** [reply, marked as answer]
Oh, I see what you mean!

Yeah, PR welcome.

**Reply by Kludex** [reply]
Honestly, I think this probably happened to me at some point as well. 😅

**Reply by scy** [reply]
@Kludex So I'm working on this PR right now, and I've noticed that middlewares don't get initialized when constructing the app. Instead, Starlette's `__call__` will call `build_middleware_stack` on first use, afaict.

Which means that we'd indeed get a 500 on the first request instead of an exception on application startup/construction.

Any ideas on how to solve this?
